### PR TITLE
UTF-8 validation

### PR DESCRIPTION
**Pull request by axic**

### CODE REVIEW

File: lib/fizzy/utf8.cpp [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    Range80BF,  // 80..BF
+    RangeA0BF,  // A0..BF
+    Range809F,  // 80..9F
+    Range90BF,  // 90..BF
+    Range808F,  // 80..8F
+};
+
+namespace fizzy
+{
+bool utf8_validate(std::string input)
```
Comment (axic) [inline lib/fizzy/utf8.cpp]: The main question I have whether this should be string or bytes_view (please also check the places this is used).

File: lib/fizzy/utf8.cpp [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    const auto len = input.size();
+
+    // Shortcut for empty inputs
+    if (len < 1)
+        return true;
+
+    unsigned required_bytes = 1;
+    auto rule = Rule::Range80BF;
+
+    const uint8_t byte1 = static_cast<uint8_t>(input[0]);
```
Comment (axic) [inline lib/fizzy/utf8.cpp]: Are you guys happy with this logic down here?

File: lib/fizzy/utf8.cpp [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+ * | U+40000..U+FFFFF   | F1..F3     | 80..BF      | 80..BF     | 80..BF      |
+ * +--------------------+------------+-------------+------------+-------------+
+ * | U+100000..U+10FFFF | F4         | 80..8F      | 80..BF     | 80..BF      |
+ * +--------------------+------------+-------------+------------+-------------+
+ */
+
+// Ranges for 2nd byte
+enum class Rule : unsigned
+{
+    Range80BF,  // 80..BF
```
Comment (axic) [inline lib/fizzy/utf8.cpp]: Could also rename the enum to `Range` and drop it from these, e.g. `Range::80BF`.

File: test/unittests/utf8_test.cpp [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+{
+    std::vector<std::pair<bytes, bool>> testcases = {
+        {"00"_bytes, true},
+        {"7f"_bytes, true},
+        {"80"_bytes, false},
+        {"c2"_bytes, false},
+        {"c280"_bytes, true},
+        {"c2bf"_bytes, true},
+        {"c2e0"_bytes, false},
+    };
```
Comment (axic) [inline test/unittests/utf8_test.cpp]: Need to add more of the edge cases here.

File: lib/fizzy/parser.cpp [hunk 5]
```diff
@@ -143,8 +144,13 @@ inline parser_result<std::string> parse_string(const uint8_t* pos, const uint8_t
     std::tie(value, pos) = parse_vec<uint8_t>(pos, end);
 
     // FIXME: need to validate that string is a valid UTF-8
+    auto const ret = std::string(value.begin(), value.end());
+    if (!utf8_validate(ret))
+    {
+        throw parser_error{"Invalid UTF8"};
```
Comment (axic) [inline lib/fizzy/parser.cpp]: Need to add a basic test case for this too.

File: lib/fizzy/utf8.cpp (hunk 1 above)
Comment (axic) [inline lib/fizzy/utf8.cpp]: Will make this a constref (as it was `bytes_view`) initially.

File: lib/fizzy/parser.cpp [hunk 6]
```diff
@@ -142,9 +143,11 @@ inline parser_result<std::string> parse_string(const uint8_t* pos, const uint8_t
     std::vector<uint8_t> value;
     std::tie(value, pos) = parse_vec<uint8_t>(pos, end);
 
-    // FIXME: need to validate that string is a valid UTF-8
+    const auto ret = std::string(value.begin(), value.end());
+    if (!utf8_validate(ret))
+        throw parser_error{"Invalid UTF8"};
 
-    return {std::string(value.begin(), value.end()), pos};
+    return {ret, pos};
```
Comment (gumb0) [inline lib/fizzy/parser.cpp]:
```suggestion
    return {std::move(ret), pos};
```

File: lib/fizzy/utf8.hpp [hunk 7]
```diff
@@ -0,0 +1,8 @@
+#pragma once
+
+#include "types.hpp"
+
+namespace fizzy
+{
+bool utf8_validate(std::string input);
```
Comment (gumb0) [inline lib/fizzy/utf8.hpp]:
```suggestion
bool utf8_validate(std::string_view input);
```

File: lib/fizzy/utf8.cpp (hunk 3 above)
Comment (gumb0) [reply lib/fizzy/utf8.cpp]: Sounds good to rename, but I don't think identifier starting with a number (`80BF`) is possible

File: lib/fizzy/utf8.cpp [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+    RangeA0BF,  // A0..BF
+    Range809F,  // 80..9F
+    Range90BF,  // 90..BF
+    Range808F,  // 80..8F
+};
+
+namespace fizzy
+{
+bool utf8_validate(const std::string& input)
```
Comment (gumb0) [inline lib/fizzy/utf8.cpp]:
```suggestion
bool utf8_validate(const std::string& input) noexcept
```

File: lib/fizzy/utf8.cpp [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+        return true;
+
+    unsigned required_bytes = 1;
+    auto rule = Rule::Range80BF;
+
+    const uint8_t byte1 = static_cast<uint8_t>(input[0]);
+    if (byte1 <= 0x7F)
+    {
+        // Shortcut for valid ASCII (also valid UTF-8)
+        return true;
```
Comment (chfast) [inline lib/fizzy/utf8.cpp]: This is nice.

File: lib/fizzy/utf8.hpp [hunk 10]
```diff
@@ -0,0 +1,8 @@
+#pragma once
+
+#include "types.hpp"
```
Comment (chfast) [inline lib/fizzy/utf8.hpp]: I don't think this is needed here.

File: lib/fizzy/utf8.hpp [hunk 11]
```diff
@@ -0,0 +1,8 @@
...
+
+namespace fizzy
+{
+bool utf8_validate(const std::string& input);
```
Comment (chfast) [inline lib/fizzy/utf8.hpp]:
```suggestion
bool utf8_validate(std::string_view input);
```
To be more generic.

File: lib/fizzy/utf8.cpp (hunk 3 above)
Comment (chfast) [reply lib/fizzy/utf8.cpp]: You cannot use identifier `80BF`.

Comment (axic) [reply lib/fizzy/utf8.cpp]: Yeah realised that after.

File: lib/fizzy/utf8.hpp (hunk 10 above)
Comment (axic) [reply lib/fizzy/utf8.hpp]: It was, for `bytes_view`.

**Comment by axic** [comment]
@chfast @gumb0 are we decided to have `utf8_validate` operate on `std::string_view`? If yes, I'll clean it up.

File: lib/fizzy/utf8.cpp [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+        if (byte2 < 0x90 || byte2 > 0xBF)
+            return false;
+        break;
+    case Rule::Range808F:
+        if (byte2 < 0x80 || byte2 > 0x8F)
+            return false;
+        break;
+    }
+
+    // Byte3 and Byte4 always have regular encoded
```
Comment (gumb0) [inline lib/fizzy/utf8.cpp]:
```suggestion
    // Byte3 and Byte4 always have regular encoding
```

File: lib/fizzy/utf8.cpp (hunk 2 above)
Comment (gumb0) [reply lib/fizzy/utf8.cpp]: Looks ok to me.

Comment (gumb0) [reply lib/fizzy/utf8.cpp]:
```suggestion
    const auto byte1 = static_cast<uint8_t>(input[0]);
```

File: lib/fizzy/utf8.cpp (hunk 8 above)
Comment (axic) [inline lib/fizzy/utf8.cpp]: Oh damn, need to change this to actually iterate the entire string, haha

File: lib/fizzy/utf8.cpp [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+    if (required_bytes > 2)
+    {
+        const uint8_t byte3 = static_cast<uint8_t>(input[2]);
+        if (byte3 < 0x80 || byte3 > 0xbf)
```
Comment (gumb0) [inline lib/fizzy/utf8.cpp]: Having consistently either lowercase or uppercase hex literals in the function would be nice

**Comment by chfast** [comment]
> @chfast @gumb0 are we decided to have `utf8_validate` operate on `std::string_view`? If yes, I'll clean it up.

It can be either here (the `bytes` type abstraction is a bit broken anyway, I wander if there is a better way). But the final type is `string`/`string_view` because we want to use it for names comparison.

File: lib/fizzy/utf8.hpp (hunk 10 above)
Comment (chfast) [reply lib/fizzy/utf8.hpp]: About that, it would be nice to move `bytes` / `bytes_view` out of `types.hpp`. Using `types.hpp` suggests this file depends on wasm types.

File: lib/fizzy/parser.cpp (hunk 6 above)
Comment (axic) [reply lib/fizzy/parser.cpp]: clang-tidy complained that moving `const string` is useless.

File: lib/fizzy/utf8.cpp (hunk 3 above)
Comment (axic) [reply lib/fizzy/utf8.cpp]: So should we rename this to something?

File: lib/fizzy/utf8.cpp (hunk 8 above)
Comment (axic) [reply lib/fizzy/utf8.cpp]: Done.

Comment (axic) [reply lib/fizzy/utf8.cpp]: Done.

File: lib/fizzy/utf8.hpp (hunk 10 above)
Comment (axic) [reply lib/fizzy/utf8.hpp]: > About that, it would be nice to move bytes / bytes_view out of types.hpp. Using types.hpp suggests this file depends on wasm types.

Yep, I was about to do that like 5 times already :)

**Comment by axic** [comment]
Why does clang-tidy thinks this:
```
/home/builder/project/lib/fizzy/utf8.cpp:57:13: error: repeated branch in conditional chain [bugprone-branch-clone,-warnings-as-errors]
            return false;
            ^
/home/builder/project/lib/fizzy/utf8.cpp:57:25: note: end of the original
            return false;
                        ^
/home/builder/project/lib/fizzy/utf8.cpp:99:13: note: clone 1 starts here
            return false;
            ^
/home/builder/project/lib/fizzy/utf8.cpp:69:9: error: repeated branch in conditional chain [bugprone-branch-clone,-warnings-as-errors]
        {
        ^
/home/builder/project/lib/fizzy/utf8.cpp:72:10: note: end of the original
        }
         ^
/home/builder/project/lib/fizzy/utf8.cpp:79:9: note: clone 1 starts here
        {
        ^
219 warnings generated.
```

File: lib/fizzy/parser.cpp (hunk 6 above)
Comment (gumb0) [reply lib/fizzy/parser.cpp]: yes, it is, you should make it non-const first

**Comment by gumb0** [comment]
clang-tidy is complaining that some branches are duplicated code. This would fix it:
https://github.com/wasmx/fizzy/commit/fe5e431791216a22796965dcdd4880872f3ea830

But maybe it's better to suppress the warning here.

File: lib/fizzy/utf8.hpp [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+
+#include <string_view>
+
+namespace fizzy
+{
+bool utf8_validate(const uint8_t* start, const uint8_t* end) noexcept;
+
+inline bool utf8_validate(std::string_view input) noexcept
+{
+    return utf8_validate(reinterpret_cast<const uint8_t*>(input.begin()),
```
Comment (gumb0) [inline lib/fizzy/utf8.hpp]: This should work
```suggestion
    return utf8_validate(input.data(), input.data() + input.size());
```

File: lib/fizzy/utf8.cpp [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+ * | U+10000..U+3FFFF   | F0         | 90..BF      | 80..BF     | 80..BF      |
+ * +--------------------+------------+-------------+------------+-------------+
+ * | U+40000..U+FFFFF   | F1..F3     | 80..BF      | 80..BF     | 80..BF      |
+ * +--------------------+------------+-------------+------------+-------------+
+ * | U+100000..U+10FFFF | F4         | 80..8F      | 80..BF     | 80..BF      |
+ * +--------------------+------------+-------------+------------+-------------+
+ */
+
+// Ranges for 2nd byte
+enum class Rule : unsigned
```
Comment (gumb0) [inline lib/fizzy/utf8.cpp]: could leave it as default
```suggestion
enum class Rule
```

File: lib/fizzy/utf8.cpp (hunk 3 above)
Comment (gumb0) [reply lib/fizzy/utf8.cpp]: I'm think I'm fine with current names, too.

File: lib/fizzy/utf8.hpp (hunk 14 above)
Comment (axic) [reply lib/fizzy/utf8.hpp]: It doesn't, `string` uses signed chars.

File: lib/fizzy/utf8.cpp [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+    Range80BF,  // 80..BF
+    RangeA0BF,  // A0..BF
+    Range809F,  // 80..9F
+    Range90BF,  // 90..BF
+    Range808F,  // 80..8F
+};
+
+namespace fizzy
+{
+bool utf8_validate(const uint8_t* pos, const uint8_t* end) noexcept
```
Comment (gumb0) [inline lib/fizzy/utf8.cpp]: Actually random access iterators can be compared with `<`, so you could still use `std::string_view` here and use iterators inside.

File: lib/fizzy/parser.cpp (hunk 6 above)
Comment (axic) [reply lib/fizzy/parser.cpp]: Yeah, but will it not be moved automatically if i is const? That's the intuition.

Comment (gumb0) [reply lib/fizzy/parser.cpp]: on the contrary, if it's bound to const variable, this variable can't be modified, so it can't be moved out of this variable

Comment (axic) [reply lib/fizzy/parser.cpp]: Okay, fixed.

File: lib/fizzy/utf8.hpp (hunk 14 above)
Comment (chfast) [reply lib/fizzy/utf8.hpp]: Still I don't think it is safe to do it like that, i.e. `reinterpreter_cast` a `string_view::const_iterator` to `const uint8_t*`. You need something like that:
```
const char* begin = input.begin();
utf8_validate(reinterpret_cast<const uint8_t*>(begin)
```

Comment (gumb0) [reply lib/fizzy/utf8.hpp]: Or make the parameter `bytes_view input` and construct `bytes_view` over vector's data in the caller.

I don't see the necessity for `reinterpret_cast` here.

File: test/unittests/utf8_test.cpp [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+#include <test/utils/hex.hpp>
+
+using namespace fizzy;
+
+TEST(utf8, invalid_first_bytes)
+{
+    std::vector<std::string> testcases;
+    for (unsigned c = 0x80; c < 0xC2; c++)
+        testcases.push_back(std::string(1, static_cast<char>(c)));
+    for (unsigned c = 0xF5; c < 0xFF; c++)
```
Comment (gumb0) [inline test/unittests/utf8_test.cpp]:
```suggestion
    for (unsigned c = 0xF5; c <= 0xFF; c++)
```

File: lib/fizzy/utf8.hpp (hunk 14 above)
Comment (axic) [reply lib/fizzy/utf8.hpp]: Actually this helper is not needed if #196 is merged first.

File: test/unittests/utf8_test.cpp (hunk 17 above)
Comment (axic) [reply test/unittests/utf8_test.cpp]: Fixed.

**Comment by axic** [comment]
Benchmarks:

`master`:
```
parse_string/16                                           19.3 ns         18.7 ns     31809506 items_per_second=26.9608/s
parse_string/32                                           99.3 ns         97.6 ns      7200091 items_per_second=45.5362/s
parse_string/64                                            103 ns          101 ns      7043529 items_per_second=90.2893/s
parse_string/128                                           106 ns         99.3 ns      7041616 items_per_second=182.986/s
parse_string/256                                           195 ns          185 ns      4333668 items_per_second=319.245/s
parse_string/512                                           208 ns          193 ns      3697522 items_per_second=717.696/s
parse_string/1024                                          165 ns          136 ns      5082887 items_per_second=1.48389k/s
parse_string/2048                                          165 ns          156 ns      4420671 items_per_second=2.97365k/s
parse_string/4096                                          201 ns          195 ns      3518596 items_per_second=5.97977k/s
```

`utf8`:
```
parse_string/16                                           25.2 ns         24.9 ns     28422125 items_per_second=22.6448/s
parse_string/32                                            111 ns          110 ns      5943587 items_per_second=49.037/s
parse_string/64                                            128 ns          125 ns      5898611 items_per_second=87.0098/s
parse_string/128                                           176 ns          163 ns      4437429 items_per_second=177.407/s
parse_string/256                                           275 ns          261 ns      2671215 items_per_second=367.722/s
parse_string/512                                           341 ns          337 ns      1947105 items_per_second=781.252/s
parse_string/1024                                          582 ns          533 ns      1091737 items_per_second=1.75942k/s
parse_string/2048                                          858 ns          841 ns       817175 items_per_second=2.97981k/s
parse_string/4096                                         1614 ns         1590 ns       399546 items_per_second=6.44811k/s
```

**Comment by chfast** [comment]
> Running this on non-random input would be better for comparison though.

All benchmarks use ASCII-only strings.

And you should be using `compare.py` to compare results.

**Comment by chfast** [comment]
```
compare.py benchmarks bin/fizzy-bench-internal-master bin/fizzy-bench-internal --benchmark_filter=parse_string                  ✔  10s    utf8 *5   14:20:45  
RUNNING: bin/fizzy-bench-internal-master --benchmark_filter=parse_string --benchmark_out=/tmp/tmpwse1anx2
2020-02-26 14:20:57
Running bin/fizzy-bench-internal-master
Run on (8 X 4400 MHz CPU s)
CPU Caches:
  L1 Data 32K (x4)
  L1 Instruction 32K (x4)
  L2 Unified 256K (x4)
  L3 Unified 8192K (x1)
Load Average: 0.42, 0.45, 0.23
----------------------------------------------------------------------------
Benchmark                  Time             CPU   Iterations UserCounters...
----------------------------------------------------------------------------
parse_string/16         16.6 ns         16.6 ns     41801560 items_per_second=23.0173/s
parse_string/32         16.9 ns         16.9 ns     41555425 items_per_second=45.6495/s
parse_string/64         17.7 ns         17.7 ns     39626538 items_per_second=91.2836/s
parse_string/128        19.6 ns         19.6 ns     35701343 items_per_second=183.355/s
parse_string/256        21.9 ns         21.9 ns     31997977 items_per_second=365.911/s
parse_string/512        28.0 ns         28.0 ns     24826319 items_per_second=735.261/s
parse_string/1024       39.4 ns         39.4 ns     17910409 items_per_second=1.45123k/s
parse_string/2048       67.7 ns         67.7 ns     10342245 items_per_second=2.92592k/s
parse_string/4096        122 ns          122 ns      5754696 items_per_second=5.84539k/s
RUNNING: bin/fizzy-bench-internal --benchmark_filter=parse_string --benchmark_out=/tmp/tmp6p5mc47a
2020-02-26 14:21:05
Running bin/fizzy-bench-internal
Run on (8 X 4400 MHz CPU s)
CPU Caches:
  L1 Data 32K (x4)
  L1 Instruction 32K (x4)
  L2 Unified 256K (x4)
  L3 Unified 8192K (x1)
Load Average: 0.47, 0.46, 0.23
----------------------------------------------------------------------------
Benchmark                  Time             CPU   Iterations UserCounters...
----------------------------------------------------------------------------
parse_string/16         25.4 ns         25.4 ns     27558500 items_per_second=22.8391/s
parse_string/32         35.3 ns         35.3 ns     19480634 items_per_second=46.5072/s
parse_string/64         52.6 ns         52.6 ns     13295579 items_per_second=91.479/s
parse_string/128        84.6 ns         84.6 ns      8208530 items_per_second=184.296/s
parse_string/256         147 ns          147 ns      4746202 items_per_second=366.439/s
parse_string/512         271 ns          271 ns      2571422 items_per_second=733.4/s
parse_string/1024        516 ns          516 ns      1356424 items_per_second=1.46432k/s
parse_string/2048       1008 ns         1008 ns       693222 items_per_second=2.93133k/s
parse_string/4096       1995 ns         1995 ns       349553 items_per_second=5.87506k/s
Comparing bin/fizzy-bench-internal-master to bin/fizzy-bench-internal
Benchmark                           Time             CPU      Time Old      Time New       CPU Old       CPU New
----------------------------------------------------------------------------------------------------------------
parse_string/16                  +0.5286         +0.5287            17            25            17            25
parse_string/32                  +1.0938         +1.0938            17            35            17            35
parse_string/64                  +1.9741         +1.9741            18            53            18            53
parse_string/128                 +3.3271         +3.3271            20            85            20            85
parse_string/256                 +5.7322         +5.7321            22           147            22           147
parse_string/512                 +8.6789         +8.6792            28           271            28           271
parse_string/1024               +12.0861        +12.0861            39           516            39           516
parse_string/2048               +13.8918        +13.8915            68          1008            68          1008
parse_string/4096               +15.3802        +15.3799           122          1995           122          1995
```

**Comment by chfast** [comment]
Although inputs are "randomized" the stddev is small.
```
parse_string/64                58.1 ns         58.1 ns     12080373 items_per_second=91.1537/s
parse_string/64                58.2 ns         58.2 ns     12080373 items_per_second=91.0071/s
parse_string/64                58.1 ns         58.1 ns     12080373 items_per_second=91.1318/s
parse_string/64                58.1 ns         58.1 ns     12080373 items_per_second=91.1351/s
parse_string/64                58.1 ns         58.1 ns     12080373 items_per_second=91.1421/s
parse_string/64                58.1 ns         58.1 ns     12080373 items_per_second=91.1661/s
parse_string/64                58.2 ns         58.2 ns     12080373 items_per_second=91.0048/s
parse_string/64                58.1 ns         58.1 ns     12080373 items_per_second=91.1778/s
parse_string/64                58.1 ns         58.1 ns     12080373 items_per_second=91.1327/s
parse_string/64_mean           58.1 ns         58.1 ns            9 items_per_second=91.1168/s
parse_string/64_median         58.1 ns         58.1 ns            9 items_per_second=91.1351/s
parse_string/64_stddev        0.042 ns        0.041 ns            9 items_per_second=0.0647514/s
```

File: lib/fizzy/utf8.cpp [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+        const uint8_t byte2 = *pos++;
+        switch (rule)
+        {
+        case Rule::Range80BF:
+            if (byte2 < 0x80 || byte2 > 0xBF)
+                return false;
+            break;
+        case Rule::RangeA0BF:
+            if (byte2 < 0xA0 || byte2 > 0xBF)
+                return false;
```
Comment (chfast) [inline lib/fizzy/utf8.cpp]: Codecov says this is not covered.

File: lib/fizzy/utf8.cpp [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+            break;
+        case Rule::Range90BF:
+            if (byte2 < 0x90 || byte2 > 0xBF)
+                return false;
+            break;
+        case Rule::Range808F:
+            if (byte2 < 0x80 || byte2 > 0x8F)
+                return false;
+            break;
+        default:
```
Comment (chfast) [inline lib/fizzy/utf8.cpp]: Default case not needed if using `enum class` locally.

File: lib/fizzy/utf8.cpp (hunk 18 above)
Comment (axic) [reply lib/fizzy/utf8.cpp]: Fixed.

File: lib/fizzy/utf8.cpp (hunk 19 above)
Comment (axic) [reply lib/fizzy/utf8.cpp]: Fixed.

File: lib/fizzy/utf8.cpp (hunk 18 above)
Comment (chfast) [reply lib/fizzy/utf8.cpp]: Looks like still not. See my commit added.

File: lib/fizzy/utf8.cpp [hunk 20]
```diff
@@ -114,7 +114,7 @@ bool utf8_validate(const uint8_t* pos, const uint8_t* end) noexcept
             break;
         case Rule::RangeA0BF:
             if (byte2 < 0xA0 || byte2 > 0xBF)
-                return false;
+                return true;  // FIXME: Unit test missing.
```
Comment (axic) [inline lib/fizzy/utf8.cpp]: Yeah forgot this one, added another one.

Comment (axic) [reply lib/fizzy/utf8.cpp]: Fixed.

File: lib/fizzy/utf8.hpp [hunk 21]
```diff
@@ -0,0 +1,8 @@
+#pragma once
+
+#include <string_view>
```
Comment (chfast) [inline lib/fizzy/utf8.hpp]: Nit:
```suggestion
#include <cstdint>
```

File: lib/fizzy/utf8.cpp [hunk 22]
```diff
@@ -0,0 +1,156 @@
...
+};
+
+namespace fizzy
+{
+bool utf8_validate(const uint8_t* pos, const uint8_t* end) noexcept
+{
+    while (pos < end)
+    {
+        unsigned required_bytes = 1;
+        auto rule = Rule::Range80BF;
```
Comment (gumb0) [inline lib/fizzy/utf8.cpp]: just an idea for better name
```suggestion
        auto byte2_rule = Rule::Range80BF;
```

File: test/unittests/utf8_test.cpp [hunk 23]
```diff
@@ -0,0 +1,130 @@
...
+
+TEST(utf8, invalid_first_bytes)
+{
+    std::vector<bytes> testcases;
+    for (uint8_t c = 0x80; c < 0xC2; c++)
+        testcases.push_back(bytes(1, c));
+    for (uint8_t c = 0xF5; c < 0xFF; c++)
+        testcases.push_back(bytes(1, c));
+    testcases.push_back("ff"_bytes);
+    for (auto const& testcase : testcases)
```
Comment (gumb0) [inline test/unittests/utf8_test.cpp]:
```suggestion
    for (const auto& testcase : testcases)
```

File: test/unittests/utf8_test.cpp [hunk 24]
```diff
@@ -0,0 +1,130 @@
...
+        {"f4808080"_bytes, true},
+        {"f48fbfbf"_bytes, true},
+        {"f490bfbf"_bytes, false},
+        {"f48fc0bf"_bytes, false},
+        {"f48fbfc0"_bytes, false},
+
+        // Multi-character example
+        {"616263c2bfe0a080ecbabaed9fbfee8181efaa81f09081a0f1a0a081f4819f85"_bytes, true},
+    };
+    for (auto const& testcase : testcases)
```
Comment (gumb0) [inline test/unittests/utf8_test.cpp]:
```suggestion
    for (const auto& testcase : testcases)
```

File: test/unittests/utf8_test.cpp (hunk 23 above)
Comment (axic) [reply test/unittests/utf8_test.cpp]: Can you get this into the lint config?

Comment (axic) [reply test/unittests/utf8_test.cpp]: Btw copied from `leb128_test` which uses `auto const&`.

Comment (chfast) [reply test/unittests/utf8_test.cpp]: I think clang-format is just working on it: https://reviews.llvm.org/D69764.
